File: examples/example_rollback.py
#!/bin/env python
USAGE = """
USAGE: example_rollback.py <ADDRESS OF SERVER>

Argument <ADDRESS OF SERVER> is necessary.

This script demonstrates the basic usage of state rollback.
"""

from IsaREPL import Client
import json
import sys


if len(sys.argv) != 2:
    print(USAGE)
    exit(1)

addr = sys.argv[1]

c = Client(addr)

def pp(x):
    print(json.dumps(x, indent=2))
    return x

def echo_eval (src):
    print('>>> '.join(src.splitlines(True)))
    ret = c.silly_eval(src)
    return pp(ret)

echo_eval ("""
theory HHH
imports Main
begin
lemma "\<exists>x::nat. x + 1 = 2"
""")

c.record_state("S0")

echo_eval("""
    apply (rule exI[where x=666])
""")

c.record_state("S1")
print("Recorded histories:")
pp(c.silly_history())
print("rollback to state S0")
pp(c.silly_rollback ("S0"))

echo_eval("""
    by (rule exI[where x=1], auto)
end
""")

c.close()

File: clients/ocaml/examples/example_rollback.ml
(** Example: State rollback

    This script demonstrates the basic usage of state rollback.
*)

open Isa_repl

let usage = "USAGE: example_rollback.exe <ADDRESS OF SERVER>\n\
             \n\
             This script demonstrates the basic usage of state rollback."

let echo_eval client src =
  Printf.printf ">>> %s\n" (String.concat "\n>>> " (String.split_on_char '\n' src));
  let result = eval client src in
  Printf.printf "Result: %s\n\n" (Msgpck.show result);
  result

let () =
  if Array.length Sys.argv <> 2 then begin
    Printf.eprintf "%s\n" usage;
    exit 1
  end;

  let addr = Sys.argv.(1) in
  let client = create addr "HOL" in

  try
    let _ = echo_eval client {|
theory HHH
imports Main
begin
lemma "\<exists>x::nat. x + 1 = 2"
|} in

    Printf.printf "Recording state S0...\n";
    record_state client "S0";

    let _ = echo_eval client {|
    apply (rule exI[where x=666])
|} in

    Printf.printf "Recording state S1...\n";
    record_state client "S1";

    Printf.printf "\nRecorded histories:\n";
    let hist = history client in
    Printf.printf "%s\n\n" (Msgpck.show hist);

    Printf.printf "Rollback to state S0\n";
    let rollback_result = rollback client "S0" in
    Printf.printf "%s\n\n" (Msgpck.show rollback_result);

    let _ = echo_eval client {|
    by (rule exI[where x=1], auto)
end
|} in

    close client;
    Printf.printf "Done!\n"

  with
  | REPLFail msg ->
      Printf.eprintf "REPL Error: %s\n" msg;
      close client;
      exit 1
  | e ->
      Printf.eprintf "Error: %s\n" (Printexc.to_string e);
      close client;
      exit 1
